File: Commons/Action/src/main/java/com/greentree/commons/action/observer/object/AsyncEventAction.java
package com.greentree.commons.action.observer.object;

import java.util.ArrayList;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;


/** @author Arseny Latyshev */
public class AsyncEventAction<T> extends EventAction<T> {
	
	private static final long serialVersionUID = 1L;
	
	private final ExecutorService executor;
	
	public AsyncEventAction(ExecutorService executor) {
		this.executor = executor;
	}
	
	@Override
	public void event(T t) {
		final var futures = new ArrayList<Future<?>>(listeners.size());
		for(var l : listeners) {
			final var f = executor.submit(()-> {
				l.accept(t);
			});
			futures.add(f);
		}
		for(var f : futures) {
			try {
				f.get();
			}catch(InterruptedException | ExecutionException e) {
				e.printStackTrace();
			}
		}
	}
	
	
	
}

File: Commons/Action/src/main/java/com/greentree/commons/action/observer/object/AsyncEventAction.kt
package com.greentree.commons.action.observer.`object`

import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

/** @author Arseny Latyshev
 */
class AsyncEventAction<T>(private val executor: ExecutorService) : EventAction<T>() {

	override fun event(t: T) {
		val futures: ArrayList<Future<*>> = ArrayList<Future<*>>(listeners.size())
		for(l in listeners) {
			val f: Future<*> = executor.submit {
				l(t)
			}
			futures.add(f)
		}
		for(f in futures) {
			try {
				f.get()
			} catch(e: InterruptedException) {
				e.printStackTrace()
			} catch(e: ExecutionException) {
				e.printStackTrace()
			}
		}
	}

	companion object {

		private const val serialVersionUID = 1L
	}
}
